_drafts/offensive-programming.md

```diff
@@ -0,0 +1,562 @@
+that the call did not fail.
+
+```c
+void *malloc_assert(size_t n) {
+  void *p = malloc(n)
+  ASSERT(p);
+  return p;
+  ...
+}
+```
+
+In code where the calls to `malloc()` should never fail, such as RTOS structure
+allocation, request and response buffers, etc., we can use this asserted
+version.
+
+Another common issue with RTOS-based systems is a queue becoming full and
+possibly not emptying quickly enough. As with memory depletion, this isn't a
+show-stopping issue, because the system might recover, but we want to look into
+it and see if it's something we can alleviate. We can add an `ASSERT()` to
+confirm that each queue insertion succeeded, or maybe even wrap that function if
+we want.
+
+```c
+void critical_event(void) {
+  ...
+  const bool success = xQueueSend(q, &item, 1000 /* wait 1s */)
+  ASSERT(success);
+  ...
+}
+```
+
+To help with the debugging process of a queue full, I highly suggest writing a
+[Python GDB
+script]({% post_url 2019-07-02-automate-debugging-with-gdb-python-api %}) to
+dump the contents of the queue. Then, when the system is halted or you have a
+core dump allowing you to find out what events were consuming the majority of
+the space in the queue!
+
+```
+(gdb) queue_print s_event_queue
+Queue Status: 10/10 events in queue (FULL!)
+0: Addr: 0x200070c0, event: BLE_PACKET
+1: Addr: 0x200070a8, event: TICK_EVENT
+2: Addr: 0x20007088, event: BLE_PACKET
+3: Addr: 0x20007070, event: BLE_PACKET
+4: Addr: 0x20007050, event: BLE_PACKET
+5: Addr: 0x20007038, event: BLE_PACKET
+6: Addr: 0x20007018, event: BLE_PACKET
+7: Addr: 0x20007000, event: BLE_PACKET
+8: Addr: 0x20006fe0, event: BLE_PACKET
+9: Addr: 0x20006fc8, event: BLE_PACKET
+```
+
+Looks like our queue was full of packets from our comms stack and we weren't
+processing them quickly enough. We now know where the issue is and we can find a
+solution.
+
+### Software Stalls & Deadlocks
+
+Embedded devices need to respond in a reasonable amount of time to user input
+and communication packets all while being performant and not showing any signs
+of lag or stalls.
+
+I can't count the number of times while working on previous projects that slow
+flash operations caused the system to freeze for 2-3 seconds at a time, wreaking
+havoc on the user experience causing other operations in the system to time out.
+The worst part about these issues is that they often aren't brought to our
+attention until the end-users complain.
+
+To help catch these issues before pushing firmware to external users, you can
+create and configure your [task
+watchdogs]({% post_url 2020-02-18-firmware-watchdog-best-practices %})#adding-a-task-watchdog)
+to be more aggressive, set up timers to assert after a few seconds during
+potentially long operations, and make sure to set timeouts on your threading
+system calls.
+
+```c
+void timing_sensitive_task(void) {
+  const bool success = mutex_lock(&s_mutex, 1000 /* wait 1s */);
+  ASSERT(success);
+  {
+    ...
+  }
+}
```

Comment:
if we have a task watchdog, do we need the explicit timeout?

Reply:
Good point. Added something.